util/test/paratest.server

```diff
@@ -200,144 +201,150 @@ sub feed_nodes {
-                } else {
-                    systemd ("$rem_cmd 2>&1 | grep -i 'Error in paratest.client:'");
-                }
-                $pe_command = $show_all_errors
-                  # print Error lines; terminate when we reach Test Summary
-                  ? '/^\\[Error/p; /^\\[Test Summary/q'
-                  # once an Error line is seen, print it and terminate
-                  : '/^\\[Error/{p;q}';
-                $partial_errors = `sed -n '$pe_command' $logfile`;
-                if ($partial_errors) {
-                    print "\n:( $partial_errors";
-                } else {
-                    print ":)";
-                }
-                exit (0);
-            }
-
-            push @logs, $logfile;
-            shift @testdir_list;
-        }
-
-        # wait before checking for free workers
-        sleep $sleep_time;                          
-    }
-
-    # wait for everyone to finish;
-    my ($done, $dead) = (0, 0);
-    while (! (-e "PARAHALT")) {
-        # For workers that are ready, set their last-fed time to zero.
-        @readyidv = free_workers ();
-        foreach $readyid (@readyidv) {
-            $callboard[$readyid] = 0;
-
-            # This worker's termination has been noted.
-            $node = $node_list[$readyid]; # machine name to rem exec to
-            $synchfile = "$synchdir/$node.$readyid";
-            unless (-e $synchfile) {
-                printf ("Error: synch file '$synchfile' missing\n");
-                exit (7);
-            }
-            unlink $synchfile;
-        }
-
-        # Now check that all nodes have timed out.
-        ($done, $dead) = (0, 0);
-        for ($id = 0; $id < $nodeCount; $id++) {
-            if ($callboard[$id] == 0) { $done++; next; }
-            $dead++ if ($timeout > 0) && ((time() - $callboard[$id]) > $timeout);
-        }
-        last if ($done + $dead) >= $nodeCount;
-
-        sleep $sleep_time;
-    }
-
-    # Note that hung processes on remote nodes must be killed manually
-    # if we exit this loop due to one or more workers being declared dead.
-    if ($dead > 0) {
-        print "\n[Error: paratest failed to exit cleanly]\n";
-        for ($id = 0; $id < $nodeCount; $id++) {
-            print "[Error: Worker $id on node $node_list[$id] timed out.]\n"
-                if $callboard[$id] > 0;
-        }
-    }
-
-    if (-e "PARAHALT" && ($#testdir_list > 0)) {
-        print "\nExiting early due to PARAHALT file\n";
-    }
-
-    if ($#testdir_list >= 0) {
-       print $#testdir_list+1; 
-       print " directory(s) left untested: @testdir_list\n";
-    }
-
-    $endtime = `date`; chomp $endtime;
-    if ($memleaksflag) {
-        systemd("cat $logdir/tmp.*.memleaks > $memleaks");
-        systemd("rm -f $logdir/tmp.*.memleaks");
+   
+    $activate_venv_output = `$venv_check`;
+    if ($? != 0) {
+      print $activate_venv_output;
+      systemd ("echo '$activate_venv_output' >> $fin_logfile");
```

Comment:
I could minimize the diff by just calling collect_logs() here and returning rather than having the else if you think that'd be cleaner.